### PR TITLE
[manila-csi-plugin]: Implement automatic topology awareness for Manila share provisioning

### LINKED ISSUE #2219: [manila-csi-plugin]: add AZ awareness support without specifying it in SC
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature
/kind bug

**What happened**:

By default manila SC without the specified AZ creates NFS shares independently on the node AZ.
This behavior is different comparing to cinder SC:
https://github.com/kubernetes/cloud-provider-openstack/blob/52b4f44b7429c95be4f83fb2a68c5792747e9cb9/pkg/csi/cinder/controllerserver.go#L74-L83

vs

https://github.com/kubernetes/cloud-provider-openstack/blob/52b4f44b7429c95be4f83fb2a68c5792747e9cb9/pkg/csi/manila/controllerserver.go#L167-L172

If a user wants to create a manila NFS share in a particular AZ, a dedicated SC with a topology must be created.

**What you expected to happen**:

If a storage class doesn't have AZ defined the PV with a `volumeBindingMode: WaitForFirstConsumer` must be created in a zone corresponding to a pod's node AZ.

If a user wants keep the existing behavior the `ignore-volume-az` options must be defined as it's done in Cinder CSI.

**How to reproduce it**:

create a manila PV 

**Anything else we need to know?**:

A lot of Cinder CSI func logic can be shared and used in Manila CSI logic.

**Environment**:
- openstack-cloud-controller-manager(or other related binary) version: any
- OpenStack version: any
- Others: any

### PR DESCRIPTION
**Pull request by kayrus**
**What this PR does / why we need it**:

This PR allows to automatically select the corresponding Node AZ without the topology mapping.

**Which issue this PR fixes(if applicable)**:

Fixes #2219

**Release note**:

```release-note
Implement automatic topology awareness for Manila share provisioning
```

### CODE REVIEW

**Comment by kayrus** [comment]
cc @zetaab @gman0 
as for the e2e tests, I'm trying to fix them in #2254

File: pkg/csi/manila/controllerserver.go [hunk 1]
```diff
@@ -135,9 +136,26 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 
 	sizeInGiB := bytesToGiB(requestedSize)
 
+	var accessibleTopology []*csi.Topology
+	accessibleTopologyReq := req.GetAccessibilityRequirements()
+	if cs.d.withTopology && accessibleTopologyReq != nil {
+		// All requisite/preferred topologies are considered valid. Nodes from those zones are required to be able to reach the storage.
+		// The operator is responsible for making sure that provided topology keys are valid and present on the nodes of the cluster.
+		accessibleTopology = accessibleTopologyReq.GetPreferred()
```
Comment (mdbooth) [inline pkg/csi/manila/controllerserver.go:144]: Reviewers note: this was the previous behaviour.

File: pkg/csi/manila/controllerserver.go [hunk 2]
```diff
@@ -135,9 +136,26 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
+
+		// Check if "availability" parameter is already specified, if not get preferred from Topology
+		// Required, incase share AZ is different from node AZ
+		if shareOpts.AvailabilityZone == "" {
+			shareOpts.AvailabilityZone = util.GetAZFromTopology(topologyKey, accessibleTopologyReq)
```
Comment (mdbooth) [inline pkg/csi/manila/controllerserver.go:149]: Reviewers note: If AZ was unset in shareOpts, we now set it based on any topology in the request. This is new behaviour.

**Review by mdbooth — COMMENTED** [review]
Can you summarise the cases where existing behaviour would change? i.e. What are all the ways this could break somebody?

**Comment by kayrus** [comment]
> Can you summarise the cases where existing behaviour would change? i.e. What are all the ways this could break somebody?

It depends on the defined StorageClass and whether it contains or doesn't contain the `availability` parameter. In short, if the `--with-topology` is set and the `availability` is not set, then AZ is detected automatically according to Nova AZ, following the Cinder CSI logic. This is the only behavior, which will be changed.

From my point of view, if the `availability` parameter isn't maintained, then k8s operator doesn't really care about the availability zones. For rare cases, when you need a strict Manila to Nova AZ mapping, operator should set the `allowedTopology` and this behavior wasn't changed.

**Comment by kayrus** [comment]
@mdbooth, is there anything else I can clarify for you?

@jichenjc @zetaab ready for review

I'm not sure whether it's a bug or a feature. Comparing to Cinder CSI behavior, it's a bug. I'd like to make cherry-picks for releases starting from 1.25, since our prod environment needs this fix and currently we have problems migrating to 1.26. Any objections?

**Comment by mdbooth** [comment]
I want to try to write in very clear language consumable by a user what the impact is.

I would be nervous about backporting a behavioural change to a stable release. If it's really required, could it be behind a feature gate (e.g. a command line argument) in the backports? We could release note the behaviour change in the next version and drop the feature gate.

**Comment by kayrus** [comment]
> If it's really required, could it be behind a feature gate (e.g. a command line argument) in the backports? We could release note the behaviour change in the next version and drop the feature gate.

nice suggestion. I think the easiest way to implement this is in a storage class parameters.

**Comment by mdbooth** [comment]
Please excuse me while I ramp up on CSI! I will try to explain my understanding below in very simple terms to both help solidify my own understanding, and so you can hopefully see where I've misunderstood:

This example SC is taken from [the docs](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/manila-csi-plugin/using-manila-csi-plugin.md):
```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: cephfs-gold
provisioner: cephfs.manila.csi.openstack.org
parameters:
  availability: zone-a
  ...
allowedTopologies:
  - matchLabelExpressions:
    - key: topology.manila.csi.openstack.org/zone
      values:
        - nova-1
        - nova-2
```

A volume with this storage class will be created in AZ `zone-a`, and can be attached to any node with labels matching `allowedTopologies`.

Cinder is identical in all relevant respects.

If we set `waitForFirstConsumer`, the create request will additionally have the topology of the node which wants the attachment.

In cinder, if `availability` is unset on the storage class we extract the first matching cinder topology label from the Node and set the availability zone to that. We assume that this is compatible with `allowedTopologies` on the storage class because... otherwise the pod wouldn't have been scheduled there?

In manila, if `availability` was unset we let manila pick.

With this change, if `availability` is unset we used the topology from the target node provided in the create request.

IIUC, this is almost certainly what we want. Who might be broken by this?

Anybody with mislabelled nodes who hadn't noticed because manila always used the default AZ will now see volume creation failures.
Anybody who had become used to the existing behaviour and provisioned sufficient capacity only in the default AZ.

Remediation:
* Explicitly set availability in the storage class
* Explicitly set manila node topology labels to point to the default AZ for all nodes

I'm obviously no expert in this area (working on it), but this sounds reasonable to me. I think the feature gate in the storage class is appropriate. After merging and backporting this fix we should change the default of the feature gate to true and mark it deprecated in master.

**Comment by kayrus** [comment]
> In cinder, if availability is unset on the storage class we extract the first matching cinder topology label from the Node and set the availability zone to that. We assume that this is compatible with allowedTopologies on the storage class because... otherwise the pod wouldn't have been scheduled there?

correct. the same logic now implemented in Manila CSI with a new extra `autoTopology` parameter.

> IIUC, this is almost certainly what we want. Who might be broken by this?
> Anybody with mislabelled nodes who hadn't noticed because manila always used the default AZ will now see volume creation failures.

Correct, if `topology-aware` was enabled and storage classes were not maintained enough, there was a chance to see a pending volume with new change.

> Anybody who had become used to the existing behaviour and provisioned sufficient capacity only in the default AZ.

AFAIK: if AZ is not set, manila tries to use the default AZ (basically the first AZ in the database default order, e.g. if zone-2 was added before zone-1, the zone-2 will be used). If the AZ capacity is exceeded, manila won't switch to AZ2. 

Once I introduced a new SC parameter, nodbody will be affected. The backward compatibility should remain.

**Comment by gouthampacha** [comment]
> What I'm less clear on is whether it's acceptable to change the default behaviour in a 'major' release, i.e. 1.28. In a clean installation, I'm pretty sure everybody wants the new behaviour. Ideally we would not require everybody to remember to use the featuregate flag forever to get non-broken behaviour.

another aspect of it is that this is completely opt-in. It needs the binding mode to be ``WaitForFirstConsumer`` in the storage class. Typically i've seen folks use ``Immediate`` as the binding mode with manila-csi; for example that's what is the default with storage classes created by the openshift manila csi driver operator: https://github.com/openshift/csi-driver-manila-operator/blob/8f6837f23b3d14494851cade712a01e3c242e93d/pkg/controllers/manila/manila.go#L221

File: docs/manila-csi-plugin/using-manila-csi-plugin.md [hunk 3]
```diff
@@ -50,6 +50,7 @@ Parameter | Required | Description
 `type` | _yes_ | Manila [share type](https://wiki.openstack.org/wiki/Manila/Concepts#share_type)
 `shareNetworkID` | _no_ | Manila [share network ID](https://wiki.openstack.org/wiki/Manila/Concepts#share_network)
 `availability` | _no_ | Manila availability zone of the provisioned share. If none is provided, the default Manila zone will be used. Note that this parameter is opaque to the CO and does not influence placement of workloads that will consume this share, meaning they may be scheduled onto any node of the cluster. If the specified Manila AZ is not equally accessible from all compute nodes of the cluster, use [Topology-aware dynamic provisioning](#topology-aware-dynamic-provisioning).
+`autoTopology` | _no_ | When set to "true" and the `availability` parameter is empty, the Manila CSI controller will map the Manila availability zone to a corresponding Nova availability zone.
```
Comment (jichenjc) [inline docs/manila-csi-plugin/using-manila-csi-plugin.md:53]: `the Manila CSI controller will map the Manila availability zone to a corresponding Nova availability zone.` 

is it to `a corresponding` or it's fixed since it's the target node (nova compute node)'s AZ ,so maybe state the info here? maybe point here to the new section you created..

File: examples/manila-csi-plugin/nfs/auto-topology-aware/storageclass.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+# Let's have two Manila AZs: zone-{1..2}
+# Let's have six Nova AZs: zone-{1..6}
+#
+# Manila zone-1 is accessible from nodes in zone-1 only
+# Manila zone-2 is accessible from nodes in zone-2 only
+#
+# We're provisioning into zone-1
+# availability parameter and allowedTopologies are empty, therefore the dynamic
+# share provisioning with automatic availability zone selection takes place.
+# The "volumeBindingMode" must be set to "WaitForFirstConsumer".
```
Comment (jichenjc) [inline examples/manila-csi-plugin/nfs/auto-topology-aware/storageclass.yaml:12]: WaitForFirstConsumer => is this new requirement of this PR ? I think AZ related change should not change storageClass behavior like this?

File: pkg/csi/manila/controllerserver.go [hunk 5]
```diff
@@ -147,7 +164,8 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		return nil, err
 	}
 
-	if err = verifyVolumeCompatibility(sizeInGiB, req, share, shareOpts, cs.d.compatOpts, shareTypeCaps); err != nil {
+	err = verifyVolumeCompatibility(sizeInGiB, req, share, shareOpts, cs.d.compatOpts, shareTypeCaps)
```
Comment (jichenjc) [inline pkg/csi/manila/controllerserver.go:167]: nit: seems identical, curious why change this :)

File: pkg/util/util.go [hunk 6]
```diff
@@ -100,3 +102,28 @@ func PatchService(ctx context.Context, client clientset.Interface, cur, mod *v1.
 
 	return nil
 }
+
+func GetAZFromTopology(topologyKey string, requirement *csi.TopologyRequirement) string {
+	var zone string
+	var exists bool
+
+	defer func() { klog.V(4).Infof("detected AZ from the topology: %s", zone) }()
```
Comment (jichenjc) [inline pkg/util/util.go:110]: I think the zone log better to be V(1) so we know what's selected for trouble shooting

File: pkg/csi/manila/controllerserver.go (hunk 5 above)
Comment (kayrus) [reply pkg/csi/manila/controllerserver.go:167]: shorter lines, therefore cleaner code ;)

File: pkg/util/util.go (hunk 6 above)
Comment (kayrus) [reply pkg/util/util.go:110]: will do

File: examples/manila-csi-plugin/nfs/auto-topology-aware/storageclass.yaml (hunk 4 above)
Comment (kayrus) [reply examples/manila-csi-plugin/nfs/auto-topology-aware/storageclass.yaml:12]: I'd say it's not a requirement for this PR, but for automatic topology awareness behavior, therefore this should should reside in the storage class parameters. Do you suggest to adjust this sentence?

File: docs/manila-csi-plugin/using-manila-csi-plugin.md (hunk 3 above)
Comment (kayrus) [reply docs/manila-csi-plugin/using-manila-csi-plugin.md:53]: I can rephrase this to a "target node".

File: docs/manila-csi-plugin/using-manila-csi-plugin.md [hunk 7]
```diff
@@ -130,6 +131,32 @@ Storage AZ does not influence
           Shares in zone-a are accessible only from nodes in nova-1 and nova-2.
 ```
 
+In cases when the Manila availability zone must correspond to the Nova
+availability zone, you can set the `autoTopology: "true"` and omit the
+`availability` parameter. By doing so, the share will be provisioned in the
```
[the remark this answers is not in the document]
Comment (kayrus) [reply docs/manila-csi-plugin/using-manila-csi-plugin.md:136]: I can address the "WaitForFirstConsumer" here. However I haven't tested the "Immediate" behavior, will try today.

File: docs/manila-csi-plugin/using-manila-csi-plugin.md (hunk 3 above)
Comment (kayrus) [reply docs/manila-csi-plugin/using-manila-csi-plugin.md:53]: addressed

**Comment by kayrus** [comment]
I addressed the requests to this PR, let me know if there is anything else left and whether we can proceed further.

File: examples/manila-csi-plugin/nfs/auto-topology-aware/storageclass.yaml (hunk 4 above)
Comment (jichenjc) [reply examples/manila-csi-plugin/nfs/auto-topology-aware/storageclass.yaml:12]: no, just want to be clear on the reason of this change :)
as long as it's reasonable I am fine with it

**Comment by jichenjc** [comment]
/approve

leave lgtm to other folks to take a look especially the comments 
https://github.com/kubernetes/cloud-provider-openstack/pull/2255#issuecomment-1563083078

**Comment by JoelSpeed** [comment]
> What I'm less clear on is whether it's acceptable to change the default behaviour in a 'major' release, i.e. 1.28. In a clean installation, I'm pretty sure everybody wants the new behaviour. Ideally we would not require everybody to remember to use the featuregate flag forever to get non-broken behaviour.

I think the change in behaviour is acceptable if you do it in a major release. You would need to make a mention of it as a bug fix in the release notes and explain the workaround for anyone who had been relying on the broken behaviour.

Proving a feature flag that allows people to preserve the old behaviour for a number of releases (normally 3?) would be preferable to prevent them being blocked on upgrades but give them a suitable timeframe (a year) to resolve the issues on their side.

Again, the feature flag and how to disable it would want to be documented in the release notes.

**Comment by kayrus** [comment]
> I think the change in behaviour is acceptable if you do it in a major release.

Do you mean that for a major release we should set `autoTopology` to true by default? And for backward compatibility a user should set it explicitly to "false"? Do you insist on specifying a CLI argument instead of SC parameter? FYI: SC parameters cannot be updated, a SC must be recreated for that.

**Comment by JoelSpeed** [comment]
> Do you mean that for a major release we should set autoTopology to true by default? And for backward compatibility a user should set it explicitly to "false"? 

I haven't dug into the depth of the PR to look at the mechanics, but, yes, if `autoToplogy` is your feature gate, then you would normally flip that from false to true in a 'major' release, so between 1.27 and 1.28 for example, but you wouldn't change it inline for a z-stream, eg, 1.27.2 to 1.27.3 would not be an acceptable way to change that default.

> Do you insist on specifying a CLI argument instead of SC parameter? FYI: SC parameters cannot be updated, a SC must be recreated for that.

Typically feature flags are implemented as a cluster wide thing rather than on the level of a resource, having different options for different SCs would probably be weird. I would have expect the feature flag default to be set via a CLI flag, those who need to retain old behaviour can do so across a single cluster, resolve any compatibility issues (by applying the correct details for their topology) and then flip the default right?

**Comment by kayrus** [comment]
@JoelSpeed for me modifying the storage class is more convenient way of the feature toggling and I'd keep it in SC. Enabling the auto topology by default for the next 1.28 release sounds reasonable. I'd prefer to get at lest one more opinion before doing this.

**Comment by mdbooth** [comment]
> @JoelSpeed for me modifying the storage class is more convenient way of the feature toggling and I'd keep it in SC. Enabling the auto topology by default for the next 1.28 release sounds reasonable. I'd prefer to get at least one more opinion before doing this.

I think this is key: what's the least disruptive way in practise to enable this behaviour change in an existing cluster? If that's changing the SC then I think the proposed design is good.

However, I thought StorageClass was immutable (but please correct me on this!). @kayrus would you be ok recreating all your storage classes to enable this feature?

**Comment by mdbooth** [comment]
Testing, testing, 1... 2... 3..., testing. Monday, Tuesday, Wednesday.

/cherry-pick release-1.27

**Comment by mdbooth** [comment]
> @mdbooth: once the present PR merges, I will cherry-pick it on top of release-1.27 in a new PR and assign it to you.

And I will assign it to @kayrus 😂

**Comment by kayrus** [comment]
> However, I thought StorageClass was immutable (but please correct me on this!).

yes, it is immutable.

>  would you be ok recreating all your storage classes to enable this feature?

for old releases I'd change the default manila storage class. Looks like there is a way to do this even though there are existing PVC's.

**Comment by JoelSpeed** [comment]
> @JoelSpeed for me modifying the storage class is more convenient way of the feature toggling and I'd keep it in SC. Enabling the auto topology by default for the next 1.28 release sounds reasonable. I'd prefer to get at least one more opinion before doing this.

I'll defer to your expertise in this area, I'm not super familiar with storage class and the relations around it

**Comment by mdbooth** [comment]
> > However, I thought StorageClass was immutable (but please correct me on this!).
> 
> yes, it is immutable.
> 
> > would you be ok recreating all your storage classes to enable this feature?
> 
> for old releases I'd change the default manila storage class. Looks like there is a way to do this even though there are existing PVC's.
> 
> > /cherry-pick release-1.27
> 
> if we agree to set autoTopology to `true` by default for 1.28, then cherry-pick for this PR is not an option.

We can't set the default to true in this PR: it will have to be a follow-on. We should do that in any case.

**Comment by mdbooth** [comment]
I think this is good. Adding a hold in case @kayrus wants to make additional changes from the above discussion. Please drop it if you're happy.

/lgtm
/hold for @kayrus

**Comment by kayrus** [comment]
I'm fine with what we already have.
/unhold
